File: web/src/store/tabs.ts
import { create } from 'zustand'
import { authedFetch } from './auth'

export type TabType = 'session' | 'project' | 'report' | 'repeating_task'

export interface Tab {
  itemType: TabType
  itemId: string
  /** ISO timestamp; updated server-side every time the tab is opened.
   *  Frontend computes the unread badge by comparing this against the
   *  source's own activity timestamp (session.last_activity, etc.). */
  lastActive: string
  /** Denormalized name of the underlying item (session/project name,
   *  repeating task name, or report title). The server resolves this
   *  on the way out — see /api/me/tabs in src/routes/me.rs. Worker
   *  sessions (`is_worker=true`) are not in the regular sessions list,
   *  so the strip relied on this field to label them; without it the
   *  chip rendered as a generic "Session" and the cleanup loop closed
   *  the tab as soon as the sessions list loaded. */
  name: string
  /** Denormalized `sessions.is_worker` for session tabs (always false
   *  for project / report / repeating-task tabs). Tabs for worker
   *  sessions hide the "Delete session" context-menu entry — worker
   *  sessions are owned by their card and the backend refuses
   *  DELETE /api/sessions/:id for them. */
  isWorker: boolean
  /** True iff the underlying session has a `repeating_task_id` set
   *  (i.e. it's a scheduled run). Tabs for these sessions hide the
   *  "Clear session" context-menu entry — clearing would wipe the run's
   *  audit trail and leave a confusing empty stub for the schedule to
   *  keep firing past. Backend enforces via POST /clear → 409. Always
   *  false for non-session tabs. */
  isRepeatingTaskSession: boolean
  /** Denormalized `sessions.is_temp` for session tabs (always false for
   *  other kinds). Temp sessions are deleted server-side — full cleanup +
   *  `session-deleted` broadcast — when their last tab (across all users)
   *  is closed. The strip marks the chip with an hourglass icon and the
   *  context menu offers "Keep session" to clear the flag. */
  isTemp: boolean
}

interface TabsState {
  tabs: Tab[]
  loaded: boolean
  fetchTabs: () => Promise<void>
  /** Open or re-activate a tab. New tabs are prepended to the strip;
   *  re-activating an existing tab does NOT change its position (tabs
   *  behave like browser tabs — clicking one selects it but doesn't
   *  shuffle the strip). Idempotent. */
  openTab: (itemType: TabType, itemId: string) => Promise<void>
  /** Remove a tab from the strip. Does not delete the underlying
   *  session/project. */
  closeTab: (itemType: TabType, itemId: string) => Promise<void>
  /** Remove a tab in response to its underlying item being deleted.
   *  Same as closeTab but kept as a separate entry point for clarity
   *  at call sites that own session/project deletion. */
  removeTabsForItem: (itemType: TabType, itemId: string) => void
  /** Reorder the strip locally by moving the tab at `fromIndex` to
   *  `toIndex` (clamped to the strip bounds). Frontend-only: the new
   *  order survives focus / poll refetches (fetchTabs preserves local
   *  order) but resets to server MRU on a full reload, since the
   *  backend has no per-tab position column. */
  moveTab: (fromIndex: number, toIndex: number) => void
}

interface ApiTab {
  item_type: TabType
  item_id: string
  last_active: string
  name?: string
  is_worker?: boolean
  is_repeating_task_session?: boolean
  is_temp?: boolean
}

function fromApi(t: ApiTab): Tab {
  return {
    itemType: t.item_type,
    itemId: t.item_id,
    lastActive: t.last_active,
    name: t.name ?? '',
    isWorker: t.is_worker ?? false,
    isRepeatingTaskSession: t.is_repeating_task_session ?? false,
    isTemp: t.is_temp ?? false,
  }
}

/** Initial sort for tabs freshly loaded from the server: most recent
 *  `lastActive` first, so a brand-new device sees the user's MRU
 *  ordering. Local order is then preserved as the user works — clicking
 *  an existing tab no longer reshuffles it. */
function sortByMru(tabs: Tab[]): Tab[] {
  return [...tabs].sort((a, b) => (a.lastActive < b.lastActive ? 1 : -1))
}

/** Build a stable key for a (type, id) pair so we can intern it in a
 *  Map. Used by the recently-closed guard below. */
function tabKey(itemType: TabType, itemId: string): string {
  return `${itemType}:${itemId}`
}

/** Tabs the user just closed (or whose underlying item was just
 *  deleted), keyed by tabKey → epoch-ms of the close. While an entry
 *  is fresh, the resurrection paths are blocked:
 *    1. an in-flight `openTab` POST response re-inserting the chip,
 *    2. a stale `fetchTabs` GET (focus / 60s poll issued before the
 *       DELETE landed server-side) re-adding it as a "new" tab.
 *  A fresh `openTab` clears the entry — closing must only stick until
 *  the user deliberately re-opens the item. Entries expire after a TTL
 *  so a DELETE that genuinely failed server-side eventually reconciles
 *  back instead of hiding the server's truth forever. Process-global
 *  (not in zustand state) because it's a transient race guard, not
 *  user-visible state. */
const recentlyClosed = new Map<string, number>()
const RECENTLY_CLOSED_TTL_MS = 30_000

function isRecentlyClosed(key: string): boolean {
  const closedAt = recentlyClosed.get(key)
  if (closedAt === undefined) return false
  if (Date.now() - closedAt > RECENTLY_CLOSED_TTL_MS) {
    recentlyClosed.delete(key)
    return false
  }
  return true
}

export const useTabsStore = create<TabsState>((set, get) => ({
  tabs: [],
  loaded: false,

  fetchTabs: async () => {
    try {
      const res = await authedFetch('/api/me/tabs')
      if (!res.ok) return
      const data = (await res.json()) as ApiTab[]
      const incoming = data
        .map(fromApi)
        // A response snapshotted before a just-issued DELETE landed
        // still contains the closed tab; without this filter the merge
        // below re-adds it as a "new" tab and the close doesn't stick.
        .filter((t) => !isRecentlyClosed(tabKey(t.itemType, t.itemId)))
      // First load: take server order verbatim (MRU). Subsequent
      // refetches (focus / 60s poll): preserve the local order the
      // user is looking at — add any new tabs at the end, drop any
      // tabs the server says are gone. Reshuffling the strip out
      // from under the user is jarring.
      set((s) => {
        if (!s.loaded) {
          return { tabs: sortByMru(incoming), loaded: true }
        }
        const incomingKey = (t: Tab) => `${t.itemType}:${t.itemId}`
        const incomingSet = new Set(incoming.map(incomingKey))
        const incomingMap = new Map(incoming.map((t) => [incomingKey(t), t]))
        const kept = s.tabs
          .filter((t) => incomingSet.has(incomingKey(t)))
          .map((t) => incomingMap.get(incomingKey(t)) ?? t)
        const keptKeys = new Set(kept.map(incomingKey))
        const added = incoming.filter((t) => !keptKeys.has(incomingKey(t)))
        return { tabs: [...kept, ...added], loaded: true }
      })
    } catch {
      // Network errors are non-fatal here — the tab strip still renders
      // whatever it has from the last successful fetch.
    }
  },

  openTab: async (itemType, itemId) => {
    // A deliberate open lifts the recently-closed guard: the user
    // clicked the item again, so storing its tab is allowed from here.
    recentlyClosed.delete(tabKey(itemType, itemId))
    const existing = get().tabs.find((t) => t.itemType === itemType && t.itemId === itemId)
    if (existing) {
      // Tab already open — selecting it should not move it. Don't
      // re-sort, don't even refresh `last_active` (we used to,
      // which then made the strip reshuffle on the next refetch).
      return
    }
    // New tab: prepend to the strip optimistically, then write through.
    // Empty `name` is fine — the upsert response (or the next refetch)
    // backfills it, and TabBar falls back to a sensible default.
    const now = new Date().toISOString()
    set((s) => ({
      tabs: [
        {
          itemType,
          itemId,
          lastActive: now,
          name: '',
          isWorker: false,
          isRepeatingTaskSession: false,
          isTemp: false,
        },
        ...s.tabs,
      ],
    }))

    try {
      const res = await authedFetch('/api/me/tabs', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ item_type: itemType, item_id: itemId }),
      })
      if (res.ok) {
        const tab = fromApi((await res.json()) as ApiTab)
        // Replace the optimistic entry in place. If a concurrent
        // `fetchTabs` happened to wipe it before our POST returned
        // (race during initial auth), re-insert at the front so the
        // tab isn't silently lost. EXCEPT when the wipe came from the
        // user closing the tab (`closeTab`) or the item being deleted
        // (`removeTabsForItem`) while our POST was in flight — the
        // recently-closed guard tells us not to resurrect the chip.
        const key = tabKey(tab.itemType, tab.itemId)
        const tombstoned = isRecentlyClosed(key)
        set((s) => {
          const exists = s.tabs.some((t) => t.itemType === tab.itemType && t.itemId === tab.itemId)
          if (exists) {
            return {
              tabs: s.tabs.map((t) =>
                t.itemType === tab.itemType && t.itemId === tab.itemId ? tab : t,
              ),
            }
          }
          // Disappeared from the strip while we were waiting on the
          // POST. If it was tombstoned (deleted-elsewhere), leave it
          // out; otherwise restore (initial-auth fetchTabs race).
          return tombstoned ? {} : { tabs: [tab, ...s.tabs] }
        })
      } else {
        // Server refused (404 = referenced item is gone, 4xx more
        // generally). Roll the optimistic add back so the strip
        // doesn't leak a phantom chip for an item the server won't
        // store.
        set((s) => ({
          tabs: s.tabs.filter((t) => !(t.itemType === itemType && t.itemId === itemId)),
        }))
      }
    } catch {
      // Leave the optimistic update in place; next focus refetch will
      // reconcile.
    }
  },

  closeTab: async (itemType, itemId) => {
    // Guard before the optimistic remove so an in-flight `openTab`
    // POST or a stale `fetchTabs` GET can't re-insert the chip. Only a
    // fresh `openTab` (the user clicking the item again) lifts it.
    recentlyClosed.set(tabKey(itemType, itemId), Date.now())
    set((s) => ({
      tabs: s.tabs.filter((t) => !(t.itemType === itemType && t.itemId === itemId)),
    }))
    try {
      // Report ids embed a `/` (`<folder>/<file>`); unencoded it adds
      // a path segment, `/api/me/tabs/{item_type}/{item_id}` never
      // matches, and the server keeps the row — i.e. a tab that won't
      // stay closed. Encoding is a no-op for UUID ids.
      await authedFetch(`/api/me/tabs/${itemType}/${encodeURIComponent(itemId)}`, {
        method: 'DELETE',
      })
    } catch {
      // ditto — focus refetch reconciles
    }
  },

  removeTabsForItem: (itemType, itemId) => {
    // Guard the id so an in-flight `openTab` POST response or a stale
    // `fetchTabs` GET can't re-insert the chip after we drop it here.
    recentlyClosed.set(tabKey(itemType, itemId), Date.now())
    set((s) => ({
      tabs: s.tabs.filter((t) => !(t.itemType === itemType && t.itemId === itemId)),
    }))
  },

  moveTab: (fromIndex, toIndex) => {
    set((s) => {
      if (fromIndex < 0 || fromIndex >= s.tabs.length) return {}
      const to = Math.max(0, Math.min(toIndex, s.tabs.length - 1))
      if (to === fromIndex) return {}
      const next = [...s.tabs]
      const [moved] = next.splice(fromIndex, 1)
      next.splice(to, 0, moved)
      return { tabs: next }
    })
  },
}))

/** Set up cross-device sync: re-fetch when the window regains focus,
 *  and on a 60s timer as a fallback for never-focused tabs. */
export function startTabsAutoSync(): () => void {
  const refresh = () => useTabsStore.getState().fetchTabs()
  const onVis = () => {
    if (document.visibilityState === 'visible') refresh()
  }
  window.addEventListener('focus', refresh)
  document.addEventListener('visibilitychange', onVis)
  const interval = window.setInterval(refresh, 60_000)
  return () => {
    window.removeEventListener('focus', refresh)
    document.removeEventListener('visibilitychange', onVis)
    window.clearInterval(interval)
  }
}

File: src/db/crud/tabs.rs
use diesel::prelude::*;

use crate::db::Db;
use crate::db::models::*;
use crate::db::schema::*;

impl Db {
    /// List a user's tabs in most-recently-active first order.
    pub async fn list_user_tabs(&self, user_id: &str) -> anyhow::Result<Vec<UserTab>> {
        let user_id = user_id.to_string();
        self.with_conn(move |conn| {
            user_tabs::table
                .filter(user_tabs::user_id.eq(&user_id))
                .select(UserTab::as_select())
                .order(user_tabs::last_active.desc())
                .load(conn)
                .map_err(Into::into)
        })
        .await
    }

    /// Insert a tab, or bump its `last_active` if it already exists.
    /// This is what powers MRU ordering: every time the user opens a
    /// session/project/report/repeating_task, the corresponding tab
    /// moves to the front.
    ///
    /// Returns `Ok(None)` if the referenced item does not exist (or
    /// `item_type` is unknown). `user_tabs` is polymorphic so there's
    /// no FK to lean on; without this check a stale URL like
    /// `/sessions/<deleted-id>` (or a cross-device delete race) would
    /// write an orphan row that then renders as a phantom chip.
    ///
    /// For DB-backed kinds (session/project/repeating_task) the
    /// existence check is a `SELECT id` against the owning table. The
    /// file-backed `report` kind has no DB row to check; callers must
    /// pre-validate the report file exists before calling.
    pub async fn upsert_user_tab(
        &self,
        user_id: &str,
        item_type: &str,
        item_id: &str,
    ) -> anyhow::Result<Option<UserTab>> {
        let now = chrono::Utc::now().to_rfc3339();
        let tab = UserTab {
            user_id: user_id.to_string(),
            item_type: item_type.to_string(),
            item_id: item_id.to_string(),
            last_active: now,
        };
        self.with_conn(move |conn| {
            let exists = match tab.item_type.as_str() {
                "session" => sessions::table
                    .find(&tab.item_id)
                    .select(sessions::id)
                    .first::<String>(conn)
                    .optional()?
                    .is_some(),
                "project" => projects::table
                    .find(&tab.item_id)
                    .select(projects::id)
                    .first::<String>(conn)
                    .optional()?
                    .is_some(),
                "repeating_task" => repeating_tasks::table
                    .find(&tab.item_id)
                    .select(repeating_tasks::id)
                    .first::<String>(conn)
                    .optional()?
                    .is_some(),
                // Reports live on disk, not in the DB. The caller validates
                // existence; the DB layer trusts the route to have done so.
                "report" => true,
                _ => false,
            };
            if !exists {
                return Ok::<Option<UserTab>, anyhow::Error>(None);
            }
            let row = diesel::insert_into(user_tabs::table)
                .values(&tab)
                .on_conflict((user_tabs::user_id, user_tabs::item_type, user_tabs::item_id))
                .do_update()
                .set(user_tabs::last_active.eq(&tab.last_active))
                .returning(UserTab::as_returning())
                .get_result(conn)?;
            Ok(Some(row))
        })
        .await
    }

    pub async fn delete_user_tab(
        &self,
        user_id: &str,
        item_type: &str,
        item_id: &str,
    ) -> anyhow::Result<bool> {
        let user_id = user_id.to_string();
        let item_type = item_type.to_string();
        let item_id = item_id.to_string();
        self.with_conn(move |conn| {
            let count = diesel::delete(
                user_tabs::table
                    .filter(user_tabs::user_id.eq(&user_id))
                    .filter(user_tabs::item_type.eq(&item_type))
                    .filter(user_tabs::item_id.eq(&item_id)),
            )
            .execute(conn)?;
            Ok(count > 0)
        })
        .await
    }

    /// Drop every user_tabs row pointing at the given (item_type, item_id),
    /// across all users. Used by cascade deletes for polymorphic tab
    /// kinds (session/project/repeating_task) so the strip doesn't
    /// render orphan chips after a delete. Reports are file-backed and
    /// have no delete path, so they don't go through this helper.
    pub async fn delete_user_tabs_for_item(
        &self,
        item_type: &str,
        item_id: &str,
    ) -> anyhow::Result<usize> {
        let item_type = item_type.to_string();
        let item_id = item_id.to_string();
        self.with_conn(move |conn| {
            let count = diesel::delete(
                user_tabs::table
                    .filter(user_tabs::item_type.eq(&item_type))
                    .filter(user_tabs::item_id.eq(&item_id)),
            )
            .execute(conn)?;
            Ok(count)
        })
        .await
    }

    /// Count `user_tabs` rows (across ALL users) pointing at the given
    /// (item_type, item_id). Powers the temp-session rule: the session is
    /// deleted only when the LAST tab pointing at it goes away, so another
    /// user who still has the tab open keeps the session alive.
    pub async fn count_user_tabs_for_item(
        &self,
        item_type: &str,
        item_id: &str,
    ) -> anyhow::Result<i64> {
        let item_type = item_type.to_string();
        let item_id = item_id.to_string();
        self.with_conn(move |conn| {
            user_tabs::table
                .filter(user_tabs::item_type.eq(&item_type))
                .filter(user_tabs::item_id.eq(&item_id))
                .count()
                .get_result(conn)
                .map_err(Into::into)
        })
        .await
    }
}
